proposals/low-level-struct-improvements.md

```diff
@@ -2,46 +2,41 @@ Low Level Struct Improvements
 =====
 
 ## Summary
-This proposal is an aggregation of several different proposals for `struct` performance improvements: ref fields and attributes to override lifetime defaults.  The goal being a design which takes into account the various proposals to create a single overarching feature set for `struct` improvements.
+This proposal is an aggregation of several different proposals for `struct` performance improvements: `ref` fields and the ability to override lifetime defaults. The goal being a design which takes into account the various proposals to create a single overarching feature set for low level `struct` improvements.
 
 ## Motivation
-features to the language: `ref` returns, `ref struct`, function pointers, etc. ... These enabled .NET developers to create write highly performant code while continuing to leverage the C# language rules for type and memory safety.  It also allowed the creation of fundamental performance types in the .NET libraries like `Span<T>`.
+Earlier versions of C# added a number of low level performance features to the language: `ref` returns, `ref struct`, function pointers, etc. ... These enabled .NET developers to create write highly performant code while continuing to leverage the C# language rules for type and memory safety.  It also allowed the creation of fundamental performance types in the .NET libraries like `Span<T>`.
 
 As these features have gained traction in the .NET ecosystem developers, both internal and external, have been providing us with information on remaining friction points in the ecosystem. Places where they still need to drop to `unsafe` code to get their work, or require the runtime to special case types like `Span<T>`. 
 
-This proposal aims to address many of these concerns by building on top of our existing low level features. Specifically it aims to:
+Today `Span<T>` is accomplished by using the `internal` type `ByReferenec<T>` which the runtime effectively treats as a `ref` field. This provides the benefit of `ref` fields but with the downside that the language provides no safety verification for it, as it does for other uses of `ref`. Further only dotnet/runtime can use this type as it's `internal`, so 3rd parties can not design their own primitives based on `ref` fields. Part of the [motivation for this work](https://github.com/dotnet/runtime/issues/32060) is to remove `ByReference<T>` and use proper `ref` fields in all code bases. 
+
+This proposal plans to address these issues by building on top of our existing low level features. Specifically it aims to:
 
 - Allow `ref struct` types to declare `ref` fields.
-- Allow the runtime to fully define `Span<T>` using the C# type system and 
-remove special case type like `ByReference<T>`
+- Allow the runtime to fully define `Span<T>` using the C# type system and remove special case type like `ByReference<T>`
```

Comment:
are tehre any other speical types the runtime has that they could get rid of as well now? (just curious)

Reply:
No other special types but it's likely that once we have these features we can essentially eliminate the concept of restricted types. They are expressable now in terms of a `ref struct` with a `ref` field.

Reply:
FWIW an existing special\restricted types being discussed for supporting https://github.com/dotnet/runtime/issues/45152 (`ref struct` + reflection with related "fast invoke" feature) is `TypedReference` which contains a loosely-typed `ByReference` expressed as a [`ByReference<byte>`](https://github.com/dotnet/runtime/blob/main/src/coreclr/System.Private.CoreLib/src/System/TypedReference.CoreCLR.cs#L16).

In order to support https://github.com/dotnet/runtime/issues/45152, a `static TypedReference FromRef<T>(ref T value)` [factory method](https://github.com/dotnet/runtime/compare/main...steveharter:FastInvoke2#diff-df560ce5bcf241f546b22cd124819fcf314e42ec480aaaf0c66fd653ce92e38cR72) is proposed which will create a `TypedReference` that references a `ref struct`. Today calling such a factory method is not possible with a `ref struct` since a `ref struct` type can't be used as a generic parameter, however that is a different issue; also note the method may be called "FromRefStruct" if necessary to distinguish between a `ref struct` and non-`ref-structs` so a new constraint like `where T : ref struct` can be added if necessary.

Reply:
> Today calling such a factory method is not possible with a ref struct since a ref struct type can't be used as a generic parameter, however that is a different issue; 

There is going to be a proposal soon-ish about how to make `ref struct` work with generic type arguments. I don't think we fully need that though to enable the `TypedReference` support. I think as a stop gap measure we could always just adjust the intrinsic `__makeref` to handle `ref struct`.